independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateImpl.java

```diff
@@ -280,6 +308,64 @@ public String toString() {
 
     }
 
+    class Capacity {
+
+        static final int LIMIT = 64 * 1024;
+
+        final int computed;
+        // intentionally not volatile; it's not a big deal if working with an outdated value
+        int max;
+
+        Capacity() {
+            this.computed = Math.min(computeCapacity(root.blocks.get(0)), LIMIT);
+        }
+
+        void update(int length) {
+            if (length > max) {
+                max = length < LIMIT ? length : LIMIT;
+            }
+        }
+
+        int get() {
+            return Math.max(max, computed);
+        }
+
+        private int computeCapacity(SectionBlock block) {
+            // This is a bit tricky because a template can contain a lot of dynamic parts
+            // Our approach is rather conservative, i.e. try not to overestimate/waste memory
+            int ret = 0;
+            for (TemplateNode node : block.nodes) {
```

Comment:
If this is on the hot path, perhaps we should iterate using an index instead of the iterator?

Reply:
Gotcha, thanks